Feature: Add argument substitution within strings

pakagronglb opened the pull request:
Add argument substitution within strings #4 

This PR implements argument substitution within strings (issue #4). Previously, arguments were only used as entire attribute values like `name="${my_arg}"`, but now they can be embedded within strings like `name="variable_${my_arg}"`.

- Enhanced the `_sub_args` method to detect and replace argument patterns within larger strings
- Added comprehensive tests for the new functionality
- Updated documentation in README.md with examples

Example usage
```xml
<py_trees.behaviours.Success name="prefix_${arg1}_suffix" />
```

This allows for more flexible behaviour tree configurations where arguments need to be part of a larger string.

Please let me know if anything needed to be fixed.

pakagronglb committed:
Feature: Add argument substitution within strings

Previously, arguments were only used as the entire string value, but now they can be embedded within strings like 'prefix__suffix'. This allows for more flexible behavior tree configurations.

erichlf commented on py_trees_parser/parser.py:
```diff
@@ -404,6 +404,29 @@ def _sub_args(self, args, var):
                 raise ValueError(f"Argument '{var_name}' not found in arg list")
 
+        # Support for argument substitution within strings
+        elif isinstance(var, str) and "${" in var and "}" in var:
```
I think you could replace the entire conditional with just your `elif` portion.

pakagronglb replied:
Good point!

erichlf reviewed (approved):
It looks good. I did leave a comment for improvement, but it is optional. Also, please check your submission using `ruff check --diff py_trees_parser/parser.py`.

pakagronglb commented:
Thank you so much for the feedback. Will do it now! @erichlf

erichlf commented on py_trees_parser/parser.py:
```diff
@@ -404,6 +404,29 @@ def _sub_args(self, args, var):
+            result = var
+            # Find all ${...} patterns in the string
+            start_idx = 0
+            while "${" in result[start_idx:]:
```
I think this will actually catch `my_${arg1}_${arg2}`. It might be good to test that.

pakagronglb replied:
Ill make a test file too after tinkering a bit and see if it passes

pakagronglb committed:
Refactor: Simplify _sub_args method for argument substitution

Simplified the conditional logic in _sub_args to handle both full replacement and embedded arguments with a single condition.

pakagronglb committed:
Further simplify _sub_args method

Removed the special case for full argument replacement, using a single pattern-matching approach to handle both full and embedded argument substitutions.

pakagronglb committed:
Add test for multiple embedded arguments

Added test case to verify handling of multiple embedded arguments in a single string (e.g., 'multiple__and__args').

erichlf committed:
Fix formatting

erichlf commented:
Thank you so much for the contribution.

pakagronglb commented:
I'm more than happy to help out other issues as well :) @erichlf

erichlf commented:
> I'm more than happy to help out other issues as well :) @erichlf

Yeah, feel free to contribute as you like.
